### PR TITLE
[SPARK-17970][SQL] store partition spec in metastore for data source table

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

We should follow hive table and also store partition spec in metastore for data source table.
This brings 2 benefits:
1. It's more flexible to manage the table data files, as users can use `ADD PARTITION`, `DROP PARTITION` and `RENAME PARTITION`
2. We don't need to cache all file status for data source table anymore.
## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

**Comment by yhuai** [comment]
Also cc @ericl

**Comment by ericl** [comment]
Also cc @mallman

**Comment by ericl** [comment]
Could we add a feature flag for this? Then it would be easier to write tests for backwards compatibility against old data-source tables.

**Comment by ericl** [comment]
@cloud-fan I think this is pretty close once https://github.com/apache/spark/pull/15539 is merged.

The only remaining issues I know about are (1) supporting mixed case columns, and (2) optimizing the insert command to not call repair table. But as discussed we can address these separately.

**Comment by ericl** [comment]
Looks like two failing tests in FileStreamSourceSuite

**Comment by ericl** [comment]
Ok, I think this is now ready to review. I've pushed commits that add a feature flag (consolidated with prior feature flags to make it easy to revert back to 2.0 behavior), backwards compatibility, and tests cases.

The proposal for backwards compatibility is as follows.
- Newly created datasource tables will have the table property `partitionProvider=hive` and will use the metastore for partition discovery.
- Older Spark versions will not respect the partitionProvider property, and will always discover partitions from remote storage.
- When partitionProvider is set to hive, partition related hive DDLs (e.g. SHOW PARTITIONS) will work. Otherwise, an exception will be thrown telling the user to run `msck repair table` to import the table partitions into the hive metastore (this will set partitionProvider = hive).
- When the feature flag is disabled, tables with hive partition provider will behave like legacy tables. The partition DDLs will also be disabled for all datasource tables.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Is this one public? For public one, should we deprecate it first (not just remove it)?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Should we use addPartition? (recover partition is known slow)

**Review by davies — COMMENTED** [review]
Had quick look, looks good to me overall. It's better to have another people looking at the details.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: It's only in master, so changing it should be ok here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: This seems non-trivial to implement, so I'll probably leave it to a follow-up pr.

**Comment by ericl** [comment]
Rebased

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: Why change this? `AnalyzeTableCommand.calculateTotalSize` reads a special hive property `hive.exec.stagingdir`, does data source table also need to take care of it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: Can users set/unset this special property? I guess no? and `DESC TABLE`, `SHOW CREATE TABLE` should also exclude it right?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: cc @ericl @yhuai, I have isolated these hive partition hacks in `HiveExternalCatalog`, so that `HiveExternalCatalog` can provides case preserving partition related interfaces.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala]: try it again?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala]: what's the result if we run this without `refresh table`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala]: what's the result if we run this without `refresh table`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: This is needed since the logical relation `sizeInBytes` is no longer the latest size. Previously ListingFileCatalog would compute the size, but TableFileCatalog just reads whatever size was present in the table metadata last.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: I'll remove it from `SHOW CREATE TABLE`, along with `transient_lastDdlTime`. For desc table, I think it would be misleading to not show it, since otherwise users could not debug the partition provider if needed. We can rename it to `org.apache.spark/partitionProvider` to make it more clear Spark is managing this property.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala
Comment (ericl) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala]: Will do

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala
Comment (ericl) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala]: You'll get the old cached results. I think this is kind of fair since otherwise we'd have to include the flag value in the cache key, which would be kind of odd.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: cool

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I think this ok, otherwise you'd need a lot of special logic just for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: add some comments to explain this format?

**Comment by cloud-fan** [comment]
LGTM, I added a new commit to make `partitionProviderIsHive` a field and put it in table properties in `HiveExternalCatalog`, to simplify the code a little bit. Feel free to revert it if it breaks something.

**Comment by ericl** [comment]
That LGTM as well. I pushed a small fix for the failing debug mode test.

**Comment by ericl** [comment]
```
[info] - SPARK-10562: partition by column with mixed case name *** FAILED *** (605 milliseconds)
[info]   java.lang.reflect.InvocationTargetException:
[info]   at sun.reflect.GeneratedMethodAccessor156.invoke(Unknown Source)
[info]   at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]   at java.lang.reflect.Method.invoke(Method.java:497)
[info]   at org.apache.spark.sql.hive.client.Shim_v0_13.getPartitionsByFilter(HiveShim.scala:588)
```

Odd that the last commit could cause this, maybe it's a flake? jenkins test this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Let's add a special node for running a sequence of commands. We are relying on the implementation of Union at here. Let's address this in a follow-up PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: How's the cost of `AnalyzeTableCommand.calculateTotalSize`? Also, why is `sizeInBytes` not the latest size?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: Seems we have a similar logic at https://github.com/apache/spark/pull/15515/files#diff-94fbd986b04087223f53697d4b6cab24R396. Are we recovering partitions twice?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Let's also update the doc.

**Comment by yhuai** [comment]
Looks good. I left a few questions. Let me know if you want to address them in follow-up prs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: I just double checked with `create table test_sel2 USING parquet PARTITIONED BY (fieldone, fieldtwo) AS SELECT id as fieldzero, id as fieldone, id as fieldtwo from range(100)` and it is using a different path, so the recovery is not duplicated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Yeah, this is a pretty major change for 2.1. Shall we do it in a followup once the patches for 2.1 are finalized?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: I have a follow-up pr for this. Will cc you.

**Review by ericl — APPROVED** [review]

**Comment by yhuai** [comment]
Cool. I am merging this pr to unblock other tasks.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/PartitionProviderCompatibilitySuite.scala]: I just checked the old behavior. It is different from the existing behavior in our Spark 2.0 build. Let me do a quick fix to resolve it.

**Pull request closed without merging**
